### ISSUE TITLE
docker_container doesn't (can't) easily work on Atomic and similar OS's

### ISSUE THREAD

**Issue by jduncan-rva** [ansible/ansible-modules-core#3762]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

docker_container
##### ANSIBLE VERSION

```

ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### CONFIGURATION

none
##### OS / ENVIRONMENT

ansible - Fedora 23
targets - RHEL Atomic 7.2
##### SUMMARY

docker_container doesn't work for Atomic Hosts (or any OS styled similarly)

docker_container looks like it requires docker-py to be installed on the target systems for the plugin to work. You can't (easily) do that in an OS like Atomic Host. 

I couldn't find an issue to track making this work. Since this is new to 2.1 I'm assuming it's the path forward. Is there a plan to get this to work without target-based dependencies being needed?
##### STEPS TO REPRODUCE

```
- name: create glusterfs server container
  docker_container:
    name: gluster-fs
    image: registry.access.redhat.com/rhgs3/rhgs-server-rhel7
    privileged: yes
    network_mode: host
    volumes_from: glusterdata
    volumes:
      - /mnt/brick1:/mnt/container_brick1[:z]

```
##### EXPECTED RESULTS

the play should run
##### ACTUAL RESULTS

the play errors out looking for a dependency that isn't there

```

fatal: [kube2]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - cannot import name NotFound. Try `pip install docker-py`"}

```

**Comment by bcoca** [comment]
docker-py needs to be installed on the host with access to manage containers, not the containers themselves. We don't have testing with atomic (yet), currently the module should work with most other distros and a docker install.

**Comment by jduncan-rva** [comment]
I guessed when I ran into it. The work-around is pretty straight-forward. I wanted to get it recorded in here so we can add it to the TODO.

-jduncan

**Comment by dhuette13** [comment]
I get a similar error when using docker_service:

```
fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - cannot import name NotFound. Try `pip install docker-py`"}
```

Here is the play:

```
- name: compose project
    docker_service:
      project_src: /srv/docker-test
      state: present
      build: true
```

Both my local machine and target are running Ubuntu 14.04. Same version of ansible: 2.1.0.0

**Comment by bcoca** [comment]
The modules clearly state docker-py is a requirement

**Comment by jduncan-rva** [comment]
I understand that. The issue is that, in an operating system like Atomic or CoreOS, adding that requirement is far from trivial.

**Comment by bcoca** [comment]
re implementing the module w/o docker-py is also far from trivial, specially if the tools on each platform vary in this way.

**Comment by jduncan-rva** [comment]
This is a much larger issue than just one module. 

For example, the postgresql_db module requires the psycopg2 module to be installed. The same problem would exist.

I'll ping the ansible product managers out-of-band to see if this is in the planning somewhere.

**Comment by wenottingham** [comment]
With my PM hat on, having container modules be container-runtime independent would be a worthwhile goal, although one that might be implemented by a set of different modules.

However the general idea of "module, written in python,  for configuring/operating service X does not require python bindings for service X" would be a non-goal; it's not practical from an ongoing code development and maintenance requirement. I'm picturing the horror of an Amazon moduleset that doesn't use boto - that's obviously an extreme example, but the principle applies.

**Comment by jduncan-rva** [comment]
Oh I totally agree. The problem I came into was that an OS designed only to run containers (Atomic) can't use the docker_container module. 

It has of course grown since then as it was sussed out, but we do need to figure that one out somehow.

**Comment by webdaford** [comment]
I'm seeing the same error on Ubuntu 16.04, but with the added twist of it working on two of my 16.04 machines, but not on another two.  I'm trying to figure out what's different between the machines, and I'm not making much progress.  They are all configured using the same Ansible playbooks, and all have docker-py installed.  They all have the same python etc.  Two work, two don't.  Any suggestions on what to look for so I can provide a better bug report?

Sample output using two of the machines ("tool" and "pender") to illustrate:

TASK [consulhost : Launch the Consul Docker image "consul"] ********************
ok: [tool]
fatal: [pender]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to import docker-py - No module named errors. Try `pip install docker-py`"}

**Comment by bcoca** [comment]
@webdaford unrelated to this ticket, the problem here is that he cannot install docker-py, use mailing lists and/or IRC for your issue.

**Comment by gregdek** [comment]
"I understand that. The issue is that, in an operating system like Atomic or CoreOS, adding that requirement is far from trivial."

While it's annoying that libraries are sometimes required on the target system, it's a fairly common pattern, even on CoreOS. 

Take a look at this project:

https://github.com/defunctzombie/ansible-coreos-bootstrap

They install python and pip as part of their bootstrapping process, and a lot of people are using this or similar for CoreOS.

We can discuss the philosophies of minimized OSes and their value, but the fact is that you can use Ansible to bootstrap any system you can reach via ssh, and people do it all the time.  We can advocate the use of this pattern until pip makes its way into Atomic natively.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by wied03** [comment]
@jduncan-rva - I'm running Atomic CentOS 7.20160707 and it came with docker.py. Have things changed since you filed thus bug?

**Comment by webdaford** [comment]
No, it's still happening.  I can install on one Ubuntu 16.04 machine, 
but not another.  I don't see any other information about what might be 
wrong or different between the two machines. Frustrating.

Dan

**Comment by wied03** [comment]
@webdaford - see @bcoca  response

**Comment by jduncan-rva** [comment]
wied03- almost certain docker-py was added to the base Atomic rpmtree. will confirm in the AM but I think this is OK to close.
